builder/builder.go

```diff
@@ -412,97 +426,448 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
+		return path, nil
+	}
+
+	// Check upstream-bundles
+	path = filepath.Join(getUpstreamBundlesPath(b.UpstreamVer), bundle)
+	if _, err := os.Stat(path); err == nil {
+		return path, nil
+	}
+
+	return "", errors.Errorf("Cannot find bundle %s in local or upstream bundles", bundle)
+}
+
+// isLocalBundle checks a bundle filepath is local
+func (b *Builder) isLocalBundle(path string) bool {
+	return strings.HasPrefix(path, b.LocalBundleDir)
+}
+
+type stringSet map[string]bool
+
+func getStringSetKeys(set stringSet) []string {
+	keys := make([]string, len(set))
+	i := 0
+	for k := range set {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+func getStringSetKeysSorted(set stringSet) []string {
+	keys := getStringSetKeys(set)
+	sort.Strings(keys)
+	return keys
+}
+
+func getBundleSetKeysSorted(set bundleSet) []string {
+	keys := make([]string, len(set))
+	i := 0
+	for k := range set {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+// getMixBundlesListAsSet reads in the Mix Bundles List file and returns the
+// resultant set of unique bundle names. If the "mixbundles" file does not
+// exist or is empty, an empty set is returned.
+func (b *Builder) getMixBundlesListAsSet() (stringSet, error) {
+	set := make(stringSet)
+
+	bundles, err := helpers.ReadFileAndSplit("mixbundles")
+	if os.IsNotExist(err) {
+		return set, nil
+	} else if err != nil {
+		return nil, errors.Wrap(err, "Failed to read in Mix Bundle List")
+	}
+
+	for _, bundle := range bundles {
+		set[bundle] = true
+	}
+	return set, nil
+}
+
+// getDirBundlesListAsSet reads the files in a directory and returns the
+// resultant set of unique bundle names. If the directory is empty, an empty
+// set is returned.
+func (b *Builder) getDirBundlesListAsSet(dir string) (stringSet, error) {
+	set := make(stringSet)
 
-	// Check if mix bundles dir exists
-	if _, err := os.Stat(b.BundleDir); os.IsNotExist(err) {
-		return bundleAddCount, errors.New("Mix bundles directory does not exist. Run mixer init-mix")
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to read bundles dir: %s", dir)
 	}
 
-	clrbundledir := getUpstreamBundlesPath(b.UpstreamVer)
+	for _, file := range files {
+		set[file.Name()] = true
+	}
+	return set, nil
+}
 
-	// Fetch upstream bundles if needed
+// writeMixBundleList writes the contents of a bundle set out to the Mix Bundles
+// List file. Values will be in sorted order.
+func (b *Builder) writeMixBundleList(bundleSet stringSet) error {
+	data := []byte(strings.Join(getStringSetKeysSorted(bundleSet), "\n"))
+	if err := ioutil.WriteFile("mixbundles", data, 0644); err != nil {
+		return errors.Wrap(err, "Failed to write out Mix Bundle List")
+	}
+	return nil
+}
+
+// buildMixBundleListFull adds a list of bundles to a bundle set, recursively
+// adding any bundles included by those in the list.
+func (b *Builder) buildMixBundleSetFull(bundles []string, set bundleSet) (bundleSet, error) {
```

Comment:
I've never done that in Go before, so I had to look up the syntax (slash look at your use of that in bundleset.go).

This builds, but is this what you were expecting?

```golang
// getFullBundleSet takes a set of bundle names to traverse, and returns a full
// set of recursively-parsed bundle objects.
func (b *Builder) getFullBundleSet(bundles bundleSet) (bundleSet, error) {
	// buildMixBundleListFull adds a list of bundles to a bundle set, recursively
	// adding any bundles included by those in the list.
	var buildMixBundleSetFull func(b *Builder, bundles []string, set bundleSet) (bundleSet, error)
	buildMixBundleSetFull = func(b *Builder, bundles []string, set bundleSet) (bundleSet, error) {
		for _, bName := range bundles {
			if _, exists := set[bName]; !exists {
				bundle, err := b.getBundleFromName(bName)
				if err != nil {
					return nil, err
				}
				set[bName] = bundle

				if len(bundle.DirectIncludes) > 0 {
					set, err = buildMixBundleSetFull(b, bundle.DirectIncludes, set)
					if err != nil {
						return nil, err
					}
				}
			}
		}
		return set, nil
	}

	set := make(bundleSet)
	set, err := buildMixBundleSetFull(b, getBundleSetKeys(bundles), set)
	if err != nil {
		return nil, err
	}

	return set, nil
}
```

The argument for this is, basically, "it doesn't make sense to ever call `buildMixBundleSetFull` except from `getFullBundleSet`, so don't expose it within the package", right?

Reply:
This local function does have access to the outer local variables. so you don't need to pass around things that won't change (like b). Because you can just mutate set, you can create it before the helper function and not pass it around. You can use ":=" so no redundant var declaration is needed right above it. Also save some indentation by early returning on the exists case.

Yes, it is mostly a scoping thing and the fact that the name of the function was confusing. Recursion helpers are a common thing, so if that's all it does, making it local or using a name like importantFunctionHelper.

Another reasonable route would make this (or some subset of the functionality) a method on bundleSet.

Reply:
Good to know!  That makes sense given what I know about Go scoping, but as I said, never inlined a func before. Not having to pass `set` around really simplifies things.

I went ahead and renamed the internal function, too, so it's less confusing.